src/remoteConnector.ts

```diff
@@ -771,62 +791,91 @@ export default class RemoteConnector extends Disposable {
 		}
 	}
 
-	public async checkRemoteConnectionSuccessful() {
-		const isRemoteExtensionHostRunning = async () => {
-			try {
-				// Invoke command from gitpot-remote extension to test if connection is successful
-				await vscode.commands.executeCommand('__gitpod.getGitpodRemoteLogsUri');
-				return true;
-			} catch {
-				return false;
-			}
-		};
+	private startHeartBeat(accessToken: string, connectionInfo: SSHConnectionParams) {
+		if (this.heartbeatManager) {
+			return;
+		}
 
-		const parseSSHDest = (sshDestStr: string): { user: string; hostName: string } | string => {
-			let decoded;
-			try {
-				decoded = JSON.parse(Buffer.from(sshDestStr, 'hex').toString('utf8'));
-			} catch {
-				// no-op
-			}
-			return decoded && typeof decoded.hostName === 'string' ? decoded : sshDestStr;
-		};
+		this.heartbeatManager = new HeartbeatManager(connectionInfo.gitpodHost, connectionInfo.workspaceId, connectionInfo.instanceId, accessToken, this.logger, this.telemetry);
+	}
 
+	private async onGitpodRemoteConnection() {
 		const remoteUri = vscode.workspace.workspaceFile || vscode.workspace.workspaceFolders?.[0].uri;
```

Comment:
I'm not sure about this check as it seems it can return undefine if workspace is slow and cannot load initial folders? Is it necessary?

Reply:
Nope, it's still a proposed API, but `workspaceFile` or `workspaceFolders` is always defined even if ssh connection failed like using invalid password